patches/server/1005-PreSpawnerSearchEvent.patch

```diff
@@ -0,0 +1,57 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: TreemanKing <67459602+TreemanKing@users.noreply.github.com>
+Date: Wed, 16 Aug 2023 11:53:44 +1000
+Subject: [PATCH] PreSpawnerSearchEvent
+
+
+diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+index 633500aefd515df5dadda3802b94079f75a03fa0..408d360679e2acafa8fb711b92ec851f7d7d1648 100644
+--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
++++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+@@ -46,6 +46,7 @@ public abstract class BaseSpawner {
+     public int requiredPlayerRange = 16;
+     public int spawnRange = 4;
+     private int tickDelay = 0; // Paper
++    private int searchCooldownTicks = 0; // Paper
+ 
+     public BaseSpawner() {}
+ 
+@@ -57,6 +58,16 @@ public abstract class BaseSpawner {
+     public boolean isNearPlayer(Level world, BlockPos pos) {
+         return world.hasNearbyAlivePlayerThatAffectsSpawning((double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange); // Paper - Affects Spawning API
+     }
++    // Paper Start - PreSpawnerSearchEvent (Convert ServerPlayers into BukkitPlayer)
++    public java.util.Collection<org.bukkit.entity.Player> convertServerPlayersToBukkitPlayers(java.util.List<net.minecraft.server.level.ServerPlayer> serverPlayers) {
++        java.util.Collection<org.bukkit.entity.Player> bukkitPlayers = new java.util.ArrayList<>(serverPlayers.size());
++
++        for (net.minecraft.server.level.ServerPlayer human : serverPlayers) {
++            bukkitPlayers.add(human.getBukkitEntity());
++        }
++        return bukkitPlayers;
++    }
++    // Paper End
+ 
+     public void clientTick(Level world, BlockPos pos) {
+         if (!this.isNearPlayer(world, pos)) {
+@@ -86,7 +97,20 @@ public abstract class BaseSpawner {
+         tickDelay = world.paperConfig().tickRates.mobSpawner;
+         if (tickDelay == -1) { return; } // If disabled
+         // Paper end
+-        if (this.isNearPlayer(world, pos)) {
++
++        // Paper start - PreSpawnerSearchEvent
++        if (--searchCooldownTicks > 0) return;
```

Comment:
We might want some input on whether this is the behaviour we want backoff ticks to have or if they should simply *just* skip the player search and use the "boolean" if players were nearby from the last attempt before the backoff.

Reply:
Additionally, letting that field dive into the negative numbers might not be the smartest idea ? Obviously underflow is far away given this is an int, but it is kind of bleh state nonetheless.

Reply:
I

> We might want some input on whether this is the behaviour we want backoff ticks to have or if they should simply _just_ skip the player search and use the "boolean" if players were nearby from the last attempt before the backoff.

Hmmm, that's a great point. I prefer the current functionality of it but it'd be interesting what others would have to say.